airbyte-integrations/connectors/source-salesforce/source_salesforce/api.py

```diff
@@ -0,0 +1,348 @@
+        self.start_date = start_date
+
+    def _get_standard_headers(self):
+        return {"Authorization": "Bearer {}".format(self.access_token)}
+
+    def get_streams_black_list(self) -> List[str]:
+        black_list = QUERY_RESTRICTED_SALESFORCE_OBJECTS + QUERY_INCOMPATIBLE_SALESFORCE_OBJECTS
+        if self.api_type == "REST":
+            return black_list
+        else:
+            return black_list + UNSUPPORTED_BULK_API_SALESFORCE_OBJECTS
+
+    def filter_streams(self, stream_name: str) -> bool:
+        if stream_name.endswith("ChangeEvent") or stream_name in self.get_streams_black_list():
+            return False
+        return True
+
+    def get_validated_streams(self, catalog: ConfiguredAirbyteCatalog = None):
+        salesforce_objects = self.describe()["sobjects"]
+        validated_streams = []
+        if catalog:
+            streams_for_read = [configured_stream.stream.name for configured_stream in catalog.streams]
+
+        for stream_object in salesforce_objects:
+            stream_name = stream_object["name"]
+            if catalog and stream_name not in streams_for_read:
+                continue
+            if self.filter_streams(stream_name):
+                validated_streams.append(stream_name)
+
+        return validated_streams
+
+    @default_backoff_handler(max_tries=5, factor=15)
+    def _make_request(
+        self, http_method: str, url: str, headers: dict = None, body: dict = None, stream: bool = False, params: dict = None
+    ) -> requests.models.Response:
+        if http_method == "GET":
+            resp = self.session.get(url, headers=headers, stream=stream, params=params)
+        elif http_method == "POST":
+            resp = self.session.post(url, headers=headers, data=body)
+        resp.raise_for_status()
+
+        return resp
+
+    def login(self):
+        login_url = f"https://{'test' if self.is_sandbox else 'login'}.salesforce.com/services/oauth2/token"
+        login_body = {
+            "grant_type": "refresh_token",
+            "client_id": self.client_id,
+            "client_secret": self.client_secret,
+            "refresh_token": self.refresh_token,
+        }
+
+        resp = self._make_request("POST", login_url, body=login_body, headers={"Content-Type": "application/x-www-form-urlencoded"})
+
+        auth = resp.json()
+        self.access_token = auth["access_token"]
+        self.instance_url = auth["instance_url"]
+
+    def describe(self, sobject: str = None) -> Mapping[str, Any]:
+        """Describes all objects or a specific object"""
+        headers = self._get_standard_headers()
+        endpoint = "sobjects" if not sobject else f"sobjects/{sobject}/describe"
+
+        url = f"{self.instance_url}/services/data/{self.version}/{endpoint}"
+        resp = self._make_request("GET", url, headers=headers)
+
+        return resp.json()
+
+    def generate_schema(self, stream_name: str) -> Mapping[str, Any]:
```

Comment:
we usually check in actual schemas to avoid having to dynamically retrieve them - what's the reason for not doing so here? there are too many?

Reply:
Yes, we get about 650 streams to work with. And also users can add custom fields for some streams, which we cannot foresee.